### PR TITLE
Refactoring steps-nowcast

### LINKED ISSUE #395: Refactor the blending/steps.py nowcasting code
Start with low-hanging fruit:
- clearer variable names (compare to nowcasts/steps.py refactoring)
- potentially make the monolithic function more modular.

### PR DESCRIPTION
**Pull request by sidekock**
I started with the steps nowcast because that seemed more straightforward to tackle first. Since both methods should look alike, they must both be changed. I can't figure out how to run the test locally, so I will have to wait and see what the pull request test gives me as feedback.

This is a first suggestion how to split up the code but look forward to suggestions and improvments

### CODE REVIEW

**Comment by sidekock** [comment]
I have a question about the Codacy Static Code Analysis. It sais the method _check_inputs is to complex while this has barely been touched since the last version, I moved a few extra lines here because that just makes more sense. Should I split up that part of the code to remove the error or not? 
image

**Comment by dnerini** [comment]
hi @sidekock , first of all, thanks for taking up this heroic task :-)

> can't figure out how to run the test locally

which part? to run the test locally you need to setup the test data correctly, see https://pysteps.readthedocs.io/en/stable/user_guide/example_data.html#example-data and https://pysteps.readthedocs.io/en/stable/developer_guide/test_pysteps.html

> I have a question about the Codacy Static Code Analysis. It sais the method _check_inputs is to complex while this has barely been touched since the last version, I moved a few extra lines here because that just makes more sense. Should I split up that part of the code to remove the error or not?

I agree with you, _check_inputs seems well in line with the complexity of the other methods, and feels like breaking it down into smaller pieces would somewhat overcrow your class. I silenced this issue on codacy.

File: pysteps/nowcasts/steps.py
Comment (dnerini) [inline pysteps/nowcasts/steps.py:1217]: Looks like it's about time to deprecate these arguments. Would you do the honors?

Comment (sidekock) [reply pysteps/nowcasts/steps.py:1217]: I will do this once I have also changed the blending/steps. This gives me some grasp of the old names to keep consistency. Once both are fully refactored, this will be the first thing on the list. 
I first wanted some feedback regarding the splitting of the steps function before I tackle the blending part because that code is quite a bit more complicated.

**Comment by sidekock** [comment]
@mats-knmi  @RubenImhoff  Could you take a look at this and give feedback? That way I can go on to the blending version. I don't want to start working on that one if this one is not in a finalized state.

**Comment by RubenImhoff** [comment]
@sidekock, I will try to do so asap.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.velocity_perturbation_method = kwargs.get("velocity_perturbation_method", "bps")
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.extrapolation_kwargs = kwargs.get("extrapolation_kwargs", None)
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.velocity_perturbation_kwargs = kwargs.get("velocity_perturbation_kwargs", None)
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.bandpass_filter = None
```
It is bandpass, right? 
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.velocity_pertubation_parallel = None
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.velocity_perturbation_perpendicular = None
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.precip_decomposed = None
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.mask_precip = None
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.mask_threshold = None
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.perturbation_generator = None
```
I'm not sure if this is the right name, what do you think?
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]: What should the full/long name be? Good to use longer names now that we are refactoring anyway.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]: What should the full/long name be? Good to use longer names now that we are refactoring anyway.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.extrapolation_method = kwargs.get("extrapolation_method", "semilagrangian")
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.decomposition_method = kwargs.get("decomposition_method", "fft")
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        self.precip_threshold = kwargs.get("precip_threshold", None)
```
If this works for you, too, then we should change it in the rest of the code as well.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]: Should we add a test where measure_time is True (and False)? That would probably tick the box for the test coverage.

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
        # Run the main nowcast loop
```

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]: Should this be:
```suggestion
        self.precip_mask = None
```
?

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]: Can we also rename this one?

**Review by RubenImhoff — CHANGES_REQUESTED** [review]
Hi @sidekock, super nice work! Thanks a lot for picking up this tremendous task. I like that we use a class now, it is much cleaner this way. 

I think all looks good, but I've made some suggestions to change some variable names to a longer name.  

In addition, perhaps `utils.nowcast_main_loop` also can use a refactor based on the refactor here. In this `nowcast_main_loop`, the `subtimesteps` are also used. I remember that we were discussing about this during the hackathon. Perhaps this is the time to clean that up a bit, too? I might be the most difficult/unreadable part of the code.

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py:44]: What about vp_prep? I would change that one too but dont know what the prep stands for

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: We could to this but I don't know if it is worth it, its an internal method with almost no functionality...

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: Is this a comment about all four of the shown names or only some?

Comment (sidekock) [reply pysteps/nowcasts/steps.py]: if it is about the last one: randoim_generator_precip ?

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: randoim_generator_motion?

**Comment by sidekock** [comment]
> Hi @sidekock, super nice work! Thanks a lot for picking up this tremendous task. I like that we use a class now, it is much cleaner this way.
> 
> I think all looks good, but I've made some suggestions to change some variable names to a longer name.
> 
> In addition, perhaps `utils.nowcast_main_loop` also can use a refactor based on the refactor here. In this `nowcast_main_loop`, the `subtimesteps` are also used. I remember that we were discussing about this during the hackathon. Perhaps this is the time to clean that up a bit, too? I might be the most difficult/unreadable part of the code.

I agree this might be worth looking at but the thing I am a bit worried about is the that this loop is used for all nowcasters so minor changes could affect the code in a lot of different places. I was thinking to tackle the loop in the blending first (since that only affect one place) and then transfer what I learned back. Its a bit of double work but it seems easier to debug that way what works and does not work. Any feedback on the way I split up the code?

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [reply pysteps/nowcasts/steps.py]: That works!

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
                    "parameter_parallel": self.velocity_perturbation_kwargs.get(
```

Last one, perhaps we should change these names too for consistency. Not sure if this suggestion is the best name, what do you think?

File: pysteps/nowcasts/steps.py
Comment (RubenImhoff) [inline pysteps/nowcasts/steps.py]:
```suggestion
                    "parameter_perpendicular": self.velocity_perturbation_kwargs.get(
```
Not sure if this suggestion is the best name, what do you think?

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: The problem is that the method "init_vel_noise()" expects these arguments to have a specific name. This method is used in multiple places, so another part of the code will crash if I change it. This shows the downside of working with all the dictionaries but I think at this point this is something we will have to accept... I could change all instances of this naming everywhere in the code, but I have no clue if this will break the code. If you feel we should do this, I will try. I can always revert to a previous commit

Comment (RubenImhoff) [reply pysteps/nowcasts/steps.py]: Good point. From that perspective, fine to leave this as is then.

**Review by RubenImhoff — APPROVED** [review]

**Comment by dnerini** [comment]
@sidekock Should this be merged?

**Comment by mats-knmi** [comment]
I am sorry for only commenting now. I have been busy with some other stuff. I will go over the changes and leave some input where I have it. If you don't want to hold off merging, that's fine since Ruben already reviewed this. Maybe you can then take my input along in the next steps of the refactoring

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: In stead of using kwargs here, you could also make a `StepsNowcasterConfig`, which would be a dataclass. This way you can more explicitly specify to the user which additional arguments need to be provided and you could even specify type hints if you want.

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: I am not a huge fan of extending `self` with additional variables not defined in `__init__`. This can lead to confusing situations, where you can be unsure if the class has a property or not. Generally I would say that variables that are created as the nowcast runs, should just be passed along in stead of stored "globally". Ofcourse just passing everything along resulted in the 20 variables per functions we had before, which might be even less ideal. Maybe the variables that need to be passed along can be bundled in dataclasses as well, which again makes everything more explicit.

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: I am wondering if it would be possible to pass only the "scaffolding" here (what are now the kwargs) and then pass `precip`, `velocity` and `timesteps`, when you actually run the nowcast. This way you could create this class once and reuse it for every new iteration.

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: The `state` could also be a dataclass. That way you could explicitly define all variables that are in the state.

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py:978]: The `params` could also be a dataclass

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [reply pysteps/nowcasts/steps.py]: Maybe, if I read the code correctly, you could create an empty `state` and `params` object (dataclass) in the `__init__` or at the start of `compute_forecast`. Then these objects can be passed along (or stored in self) to the methods that fill this state/params. This would group the currently existing variables a bit nicer together maybe?

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: One small stylistic thing is that, as you did with the methods, you could also make all "private" variable names start with an underscore.

**Comment by mats-knmi** [comment]
I have left my comments, let me know what you think. Again, sorry for taking so long to get back to you. Overall the code looks much more manageable now, compared to what it looked like before.

**Comment by sidekock** [comment]
> @sidekock Should this be merged?

Let me use the feedback of Mats. I will contact you once it is ready to be merged.

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: I understand the idea and that sounds very good indeed. If you need to run multiple, no need in setting up all these things. However, how would this work at runtime? I don't see how we can create a class once and then keep for multiple calls of the forecast function... Maybe its just to early in the morning :)

Comment (sidekock) [reply pysteps/nowcasts/steps.py]: So I see how we could do this but the old functionality of forecast() can not be changed. Only when people want to use the new framework more modularly would this be possible. In this case forecast would look something like this:
- init StepsNowcasterConfig
- init nowcaster (with precip, velocity, timesteps, and Config)
- nowcaster.compute_forecast()
This would enable people who run the forecast for multiple run times to define the StepsNowcasterConfig once and then loop over all times. This is the only solution I see right now but this would not be possible in the current forecast() method.

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: I will do that but would like to see your feedback on this at is is the first time I am using these type hints. Also @dnerini would this be something you would like to see in this version of pysteps? I know @mats-knmi is doing this for the v2 version but want to check if this is still within the pysteps 'style'

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: With private, do you mean everything needed under the hood and not explicitly used as an input or output? Just a check because of the quotations around private

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py:978]: I agree, and this fits nicely with your previous suggestion. That way we have a dataclass which can be modified (because some of these are only calculated at run time) and can be retrieved with a simple get

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [reply pysteps/nowcasts/steps.py]: What I meant is that it is possible to do something like the snippet below when using this class. If `filter_kwargs` were renamed to `_filter_kwargs`, you would indicate to the user that this is not allowed.

The reason I put private in quotations is that just putting a single underscore before a variable does not enforce this. A user can choose to ignore the underscore (but a linter like pylint would warn them that this is not allowed).

It is also possible in python to start the variable name with 2 underscores (`__filter_kwargs`). In this case the python interpreter actually (somewhat) enforces the privateness of this variable, by doing something called "name mangling". This might even be preferred to the single underscores (also for the private methods). Personally I use double underscores as much as possible.

See also: https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name

```
nowcaster = StepsNowcaster(...)
nowcaster.filter_kwargs = new_filter_kwargs
```

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [reply pysteps/nowcasts/steps.py]: Yes the old `forecast()` method can indeed not be updated to this new way of running the code, but maybe we can deprecate the old `forecast()` method over time? Actually this could easily be done when the xarray v2 would be released.

As for how this would work, I was thinking something like this:
```
nowcaster = StepsNowcast(timesteps, StepsNowcastConfig(...))
nowcaster.compute_forecast(precip, velocity)
```
The `__init__` could actually already take care of some things that don't need `precip` and `velocity` like retrieving the oflow method to be used.

The only thing you need to be very careful with in this case is that you reset the state properly every time `compute_forecast` is called.

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [reply pysteps/nowcasts/steps.py]: Yeah I was wondering about type hints as well in the v2 version. Currently it does not really fit in the pysteps style, but it is possible to use them in some places and not others without breaking the code.

I was hoping we could slowly roll them out to more parts of the code and eventually make pysteps completely type hinted (also adding mypy to the github actions), but this would still be very far off.

For now I think it doesn't hurt sprinkling them in there where they make sense, but it does indeed mean that the style is somewhat different between different parts of the code. I am interested what @dnerini thinks of this.

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: I am going step by step. This Config file is now added; please let me know if this is how you want me to continue or if I need to change the code. Currently, the config file is a class variable. All features from these variables are always accessed as follows: self.config.XXXX. That way, it is always clear what parameters are configuration parameters and which ones are parameters or states for the nowcaster. The next step is to add these two...

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: This is done now! I might need to add typing and dataclass and typing as new optional dependicies if these are not in base python...

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: I indeed missed this one... Currently I added it to the init. Next step is indeed to add the two suggested classes to remove the number of class variables

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: I want to pick this up if all the other changes you suggested are implemented. We should also check this with @dnerini to see if this does not break the PySteps rules. It is in the official documentation, but I rarely see it in the actual code.

**Comment by sidekock** [comment]
@mats-knmi could you look through the changes to give some feedback? I will look at the 'private' variables next week if I have time. In general, I think this approach is much cleaner compared to what we had before. We see what is changed and what type it is (state or param) and what is loaded (config, state or param) for each method. 

For the 'private' discussion, I have taken a small look at the StackOverflow page but do not yet see how I would use this at this moment. Most variables in the code are part of a dataclass so can I still make then private then?

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py:40]: Since pysteps dropped support for python 3.9:
`Optional[float]` can be written as `float | None`
`Dict[str, Any]` can be written as `dict[str, Any]` and same goes for `List`

Doing this saves some imports from the typing module; you don't have to import `Optional`, `Dict` and `List`

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [inline pysteps/nowcasts/steps.py]: The type hints in the dataclasses already help a lot. As an extra bonus these classes can then also be used as type hints for the methods that expect them.

Also, I believe this `__init__` method (or the class itself) should have a docstring indicating to the user what they are expected to provide as arguments, or is that done somewhere else currently?

```suggestion
    def __init__(self, precip, velocity, time_steps, steps_config: StepsNowcasterConfig):
```

**Comment by mats-knmi** [comment]
Looks good, I added 2 small comments on the type hinting, but other than that I think this makes it a lot nicer indeed. Regarding the private discussion: 
Every method that currently starts with an underscore could start with 2 underscores and then they could still be called from within the class, but not as easily from outside. (e.g. `_nowcast_main` could be renamed to `__nowcast_main` and then be called normally as `self.__nowcast_main()`).
Every variable that is defined in `__init__` could be defined with 2 underscores and then still accessed normally within the class, but not as easily from outside. (e.g. `self.__state = StepsNowcasterState()` and then acccess it later like `self.__state.precip_forecast`)

**Comment by sidekock** [comment]
> Looks good, I added 2 small comments on the type hinting, but other than that I think this makes it a lot nicer indeed. Regarding the private discussion: Every method that currently starts with an underscore could start with 2 underscores and then they could still be called from within the class, but not as easily from outside. (e.g. `_nowcast_main` could be renamed to `__nowcast_main` and then be called normally as `self.__nowcast_main()`). Every variable that is defined in `__init__` could be defined with 2 underscores and then still accessed normally within the class, but not as easily from outside. (e.g. `self.__state = StepsNowcasterState()` and then acccess it later like `self.__state.precip_forecast`)

Thanks for all the feedback @mats-knmi and @RubenImhoff. The new version of the code feels much more readable now! I made the last changes and will soon start on the real challenge: the blending code :)

**Comment by sidekock** [comment]
@dnerini the Static code analysis is complaining again because I changed the name of the method "__check_inputs". Would it be possible to disable this once more? That way, all tests should pass now

File: pysteps/nowcasts/steps.py
Comment (mats-knmi) [reply pysteps/nowcasts/steps.py]: I think that the second part of this comment maybe still needs to be addressed? (part of) the docstring from the `forecast` method probably needs to be copied here so that when users use the class directly they know what to do with each parameter. Possibly the docstring can be cut into 2 pieces where the part that describes `precip`, `velocity` and `timesteps` can be moved here and the part that describes all other parameters could be a docstring for the `StepsNowcasterConfig` dataclass.

**Comment by mats-knmi** [comment]
@sidekock I resolved all my comments that are fully resolved now, I left 2 open, since those are not fully resolved yet. 

I think the docstring comment is important to still address, but the one about passing `precip` and `velocity` and/or `timesteps` not at the `__init__` but at the `compute_forecast` is more of a stylistic preference.

File: pysteps/nowcasts/steps.py
Comment (sidekock) [reply pysteps/nowcasts/steps.py]: This should be done now

**Comment by sidekock** [comment]
> @sidekock I resolved all my comments that are fully resolved now, I left 2 open, since those are not fully resolved yet.
> 
> I think the docstring comment is important to still address, but the one about passing `precip` and `velocity` and/or `timesteps` not at the `__init__` but at the `compute_forecast` is more of a stylistic preference.

The new naming is now applied and the documentation is provided.

**Comment by sidekock** [comment]
@mats-knmi if all the requested changes are present, would you please approve the changes? Then I know I can start working on the blending part after the merge

**Comment by mats-knmi** [comment]
Yes they are approved. ~~I wanted to approve them in github, but I can't find the button to do this. Not sure if I am just blind or if I am not authorized to approve stuff...~~ Nevermind I found the button. It is in a completely different place as to where I am used to in gitlab.

**Review by mats-knmi — APPROVED** [review]

**Comment by sidekock** [comment]
@dnerini I think this is ready to be merged with the master branch

**Comment by dnerini** [comment]
Fantastic work! 🙌

**Pull request merged**
